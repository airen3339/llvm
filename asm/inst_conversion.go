package asm

import (
	"fmt"

	"github.com/llir/l/ir"
	"github.com/mewmew/l-tm/asm/ll/ast"
)

// --- [ Conversion instructions ] ---------------------------------------------

// ~~~ [ trunc ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstTrunc(inst ir.Instruction, old *ast.TruncInst) (*ir.InstTrunc, error) {
	i, ok := inst.(*ir.InstTrunc)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstTrunc, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ zext ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstZExt(inst ir.Instruction, old *ast.ZExtInst) (*ir.InstZExt, error) {
	i, ok := inst.(*ir.InstZExt)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstZExt, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ sext ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSExt(inst ir.Instruction, old *ast.SExtInst) (*ir.InstSExt, error) {
	i, ok := inst.(*ir.InstSExt)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSExt, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ fptrunc ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFPTrunc(inst ir.Instruction, old *ast.FPTruncInst) (*ir.InstFPTrunc, error) {
	i, ok := inst.(*ir.InstFPTrunc)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFPTrunc, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ fpext ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFPExt(inst ir.Instruction, old *ast.FPExtInst) (*ir.InstFPExt, error) {
	i, ok := inst.(*ir.InstFPExt)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFPExt, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ fptoui ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFPToUI(inst ir.Instruction, old *ast.FPToUIInst) (*ir.InstFPToUI, error) {
	i, ok := inst.(*ir.InstFPToUI)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFPToUI, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ fptosi ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFPToSI(inst ir.Instruction, old *ast.FPToSIInst) (*ir.InstFPToSI, error) {
	i, ok := inst.(*ir.InstFPToSI)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFPToSI, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ uitofp ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstUIToFP(inst ir.Instruction, old *ast.UIToFPInst) (*ir.InstUIToFP, error) {
	i, ok := inst.(*ir.InstUIToFP)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstUIToFP, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ sitofp ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSIToFP(inst ir.Instruction, old *ast.SIToFPInst) (*ir.InstSIToFP, error) {
	i, ok := inst.(*ir.InstSIToFP)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSIToFP, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ ptrtoint ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstPtrToInt(inst ir.Instruction, old *ast.PtrToIntInst) (*ir.InstPtrToInt, error) {
	i, ok := inst.(*ir.InstPtrToInt)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstPtrToInt, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ inttoptr ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstIntToPtr(inst ir.Instruction, old *ast.IntToPtrInst) (*ir.InstIntToPtr, error) {
	i, ok := inst.(*ir.InstIntToPtr)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstIntToPtr, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ bitcast ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstBitCast(inst ir.Instruction, old *ast.BitCastInst) (*ir.InstBitCast, error) {
	i, ok := inst.(*ir.InstBitCast)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstBitCast, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ addrspacecast ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstAddrSpaceCast(inst ir.Instruction, old *ast.AddrSpaceCastInst) (*ir.InstAddrSpaceCast, error) {
	i, ok := inst.(*ir.InstAddrSpaceCast)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstAddrSpaceCast, got %T", inst))
	}
	// TODO: implement
	return i, nil
}
