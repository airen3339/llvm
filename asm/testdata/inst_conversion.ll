define void @f() {
	trunc i32 321 to i8
	zext i8 123 to i32
	sext i8 -123 to i32
	fptrunc double 1.0 to float
	fpext float 2.0 to double
	fptoui double 3.0 to i32
	fptosi double -4.0 to i32
	uitofp i32 5 to double
	sitofp i32 -6 to double
	ptrtoint i8* null to i32
	inttoptr i32 1234 to i8*
	bitcast { i32, i32 } { i32 10, i32 20 } to i64
	addrspacecast i8* null to i8 addrspace(1)*
	ret void
}
