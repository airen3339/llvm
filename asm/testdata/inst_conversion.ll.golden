define i8 @trunc_1() {
; <label>:0
	%result = trunc i32 298 to i8
	ret i8 %result
}

define i8 @trunc_2() {
; <label>:0
	%result = trunc i32 298 to i8, !baz !{!"qux"}, !foo !{!"bar"}
	ret i8 %result
}

define i32 @zext_1() {
; <label>:0
	%result = zext i8 42 to i32
	ret i32 %result
}

define i32 @zext_2() {
; <label>:0
	%result = zext i8 42 to i32, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @sext_1() {
; <label>:0
	%result = sext i8 -42 to i32
	ret i32 %result
}

define i32 @sext_2() {
; <label>:0
	%result = sext i8 -42 to i32, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define float @fptrunc_1() {
; <label>:0
	%result = fptrunc double 42.0 to float
	ret float %result
}

define float @fptrunc_2() {
; <label>:0
	%result = fptrunc double 42.0 to float, !baz !{!"qux"}, !foo !{!"bar"}
	ret float %result
}

define double @fpext_1() {
; <label>:0
	%result = fpext float 42.0 to double
	ret double %result
}

define double @fpext_2() {
; <label>:0
	%result = fpext float 42.0 to double, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i32 @fptoui_1() {
; <label>:0
	%result = fptoui double 42.0 to i32
	ret i32 %result
}

define i32 @fptoui_2() {
; <label>:0
	%result = fptoui double 42.0 to i32, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @fptosi_1() {
; <label>:0
	%result = fptosi double -42.0 to i32
	ret i32 %result
}

define i32 @fptosi_2() {
; <label>:0
	%result = fptosi double -42.0 to i32, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @uitofp_1() {
; <label>:0
	%result = uitofp i32 42 to double
	ret double %result
}

define double @uitofp_2() {
; <label>:0
	%result = uitofp i32 42 to double, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @sitofp_1() {
; <label>:0
	%result = sitofp i32 -42 to double
	ret double %result
}

define double @sitofp_2() {
; <label>:0
	%result = sitofp i32 -42 to double, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i64 @ptrtoint_1() {
; <label>:0
	%result = ptrtoint i8* null to i64
	ret i64 %result
}

define i64 @ptrtoint_2(i8* %x) {
; <label>:0
	%result = ptrtoint i8* %x to i64
	ret i64 %result
}

define i64 @ptrtoint_3() {
; <label>:0
	%result = ptrtoint i64 ()* @ptrtoint_3 to i64
	ret i64 %result
}

define i64 @ptrtoint_4() {
; <label>:0
	%result = ptrtoint i8* null to i64, !baz !{!"qux"}, !foo !{!"bar"}
	ret i64 %result
}

define i8* @inttoptr_1() {
; <label>:0
	%result = inttoptr i32 0 to i8*
	ret i8* %result
}

define i8* @inttoptr_2() {
; <label>:0
	%result = inttoptr i32 0 to i8*, !baz !{!"qux"}, !foo !{!"bar"}
	ret i8* %result
}

define <4 x i8> @bitcast_1() {
; <label>:0
	%result = bitcast i32 707406378 to <4 x i8>
	ret <4 x i8> %result
}

define <4 x i8> @bitcast_2() {
; <label>:0
	%result = bitcast i32 707406378 to <4 x i8>, !baz !{!"qux"}, !foo !{!"bar"}
	ret <4 x i8> %result
}

define i8* @addrspacecast_1() {
; <label>:0
	%result = addrspacecast i8* null to i8*
	ret i8* %result
}

define i8* @addrspacecast_2() {
; <label>:0
	%result = addrspacecast i8* null to i8*, !baz !{!"qux"}, !foo !{!"bar"}
	ret i8* %result
}
